File: typescript/fibonacci.ts
// Algorithm: Fibonacci (Iterative DP)
// Description: Builds Fibonacci numbers up to n using bottom-up dynamic programming in O(n) time.
// Usage: Run with `ts-node fibonacci.ts` or compile using `tsc`.
export {};

function fibonacci(n: number): number[] {
  if (n < 0) {
    throw new Error("n must be non-negative");
  }
  const sequence: number[] = [];
  for (let i = 0; i <= n; i += 1) {
    if (i === 0) {
      sequence.push(0);
    } else if (i === 1) {
      sequence.push(1);
    } else {
      sequence.push(sequence[i - 1] + sequence[i - 2]);
    }
  }
  return sequence;
}

const limit = 10;
console.log(`Fibonacci sequence up to index ${limit}:`, fibonacci(limit));

File: java/fibonacci.java
/*
 * Algorithm: Fibonacci (Dynamic Programming)
 * Description: Generates Fibonacci sequence up to n using iterative DP; O(n) time.
 * Usage: Compile with `javac fibonacci.java` and run `java FibonacciDp`.
 */
class FibonacciDp {
    static long[] fibonacci(int n) {
        if (n < 2) {
            return new long[]{0, 1};
        }
        long[] sequence = new long[n + 1];
        sequence[0] = 0;
        sequence[1] = 1;
        for (int i = 2; i <= n; i++) {
            sequence[i] = sequence[i - 1] + sequence[i - 2];
        }
        return sequence;
    }

    public static void main(String[] args) {
        int n = 10; // Generate first 10 Fibonacci numbers.
        long[] sequence = fibonacci(n);
        System.out.print("Fibonacci sequence: ");
        for (int i = 0; i <= n; i++) {
            System.out.print(sequence[i] + " ");
        }
        System.out.println();
    }
}
